OMCompiler/SimulationRuntime/c/linearization/linearize.cpp

```diff
@@ -623,6 +628,16 @@ int linearize(DATA* data, threadData_t *threadData)
 
     FILE *fout = omc_fopen(filename.c_str(),"wb");
     assertStreamPrint(threadData,0!=fout,"Cannot open File %s",filename.c_str());
+    char* buffer;
+    if( (buffer=getcwd(NULL, 0)) == NULL) {
```

Comment:
It is convoluted, but your changes do not really make it less convoluted right? :D

I wrestled with this buffer thing for some time... somehow the garbage collector just frees it way to early or something like that. I had to do things in that specific order or it would just not work, so i did not touch it anymore. But maybe i should try to clean it up a bit.

Reply:
I guess the minimalist solution would be to remove the brackets on the first line
 if( buffer=getcwd(NULL, 0)) ;) 

Still not sure if that would help. I do prefeer @phannebohm solution

Reply:
Or we could change what is suggested to 
```
char* buffer = getcwd(NULL, 0);
 if(!buffer) {
```

Reply:
I don't think there is supposed to be a garbage collector here at all. Actually I don't fully understand what all that code is doing there?

(Assignments inside of if-conditions is an awesome thing but we might want to leave that to the experts at https://www.ioccc.org/)

Reply:
Oh, maybe you want to do a memcpy of the string?

Reply:
> 
> 
> I don't think there is supposed to be a garbage collector here at all. Actually I don't fully understand what all that code is doing there?
> 
> (Assignments inside of if-conditions is an awesome thing but we might want to leave that to the experts at https://www.ioccc.org/)

It should not be unless it is registered with Boehm. For instance, if you look in the MetaModelica runtime there are some codes like that mmc_alloc or whatever it is called

Reply:
Ok i don't know what happens here, but if i try to use `string()` on the buffer here and prepend it to filename, it works perfectly fine here. i can dump it and everything is fine. But if i store this very same variable in my data block and call it later, it somehow is corrupted and spits out garbage like PQw or something like that. If i just store the buffer as is and call it later it works however. So i thought that the buffer somehow gets freed to early.

Reply:
> But if i store this very same variable in my data block and call it later, it somehow is corrupted and spits out garbage like PQw or something like that. If i just store the buffer as is and call it later it works however

What do you mean by this Karim?

`getcwd(NULL, 0);`
is an GNU extension that is adapted by others and not a standard. So it might be better to not use it at all and just allocate the max path size and pass that to getcwd and then copy the returned string to an appropriate size buffer. 

Other things,
```
if( (buffer=getcwd(NULL, 0)) == NULL) {
      fullpath = "";
      free(buffer);
}
```
- if `buffer `is `NULL `should you not report error or exit or something similar? 
- It is also not necessary to `free(buffer)` if it is `NULL`. 
- `fullpath `is a local variable that is not used at all.

Get your coffee and revise it.

Reply:
I think i found a way prettier version. I just moved the dump one section down and now i got:
```
  char* buffer=getcwd(NULL, 0);
    if(data->modelData->runTestsuite || !(buffer)){
      infoStreamPrint(LOG_STDOUT, 0, "Linear model is created.");
    }else{
      infoStreamPrint(LOG_STDOUT, 0, "Linear model is created at %s/%s", buffer, filename.c_str());
      infoStreamPrint(LOG_STDOUT, 0, "The output format can be changed with the command line option --linearizationDumpLanguage.");
      infoStreamPrint(LOG_STDOUT, 0, "The options are: --linearizationDumpLanguage=modelica, matlab, julia, python.");
    }
    free(buffer);
```

Reply:
I will allocate with max path, i forgot about that. 

The thing about all this was, that if i left the scope of this function somehow nothing worked out the way i thought it would.

Reply:
```
    char cwd[PATH_MAX];
    if(data->modelData->runTestsuite || !(getcwd(cwd, sizeof(cwd)))){
      infoStreamPrint(LOG_STDOUT, 0, "Linear model is created.");
    }else{
      infoStreamPrint(LOG_STDOUT, 0, "Linear model is created at %s/%s", cwd, filename.c_str());
      infoStreamPrint(LOG_STDOUT, 0, "The output format can be changed with the command line option --linearizationDumpLanguage.");
      infoStreamPrint(LOG_STDOUT, 0, "The options are: --linearizationDumpLanguage=modelica, matlab, julia, python.");
    }
```

Reply:
> I will allocate with max path, i forgot about that.

That was just FYI. I guess we might be using the same thing in other places so it is up to you to decide. I like this one though.

However shouldn't you split the checks for runTestSuite and getcwd()? 
The reason is getcwd can fail for some reason and you still will ignore the error and just print out "Linear model is created." 
It can fail for permission reasons for example. I know it does not matter for "correctness" if you are able to print the cwd. But maybe it is better to let the user know.

```
char cwd[PATH_MAX];
if (data->modelData->runTestsuite)) {
    infoStreamPrint(LOG_STDOUT, 0, "Linear model is created.");
}
else {
    char* success = getcwd(cwd, sizeof(cwd));
    if(!success) {
        "Error geting pwd..."
    }
    else {
        infoStreamPrint(LOG_STDOUT, 0, "Linear model is created at %s/%s", cwd, filename.c_str());
    }
    infoStreamPrint(LOG_STDOUT, 0, "The output format can be changed with the command line option --linearizationDumpLanguage.");
    infoStreamPrint(LOG_STDOUT, 0, "The options are: --linearizationDumpLanguage=modelica, matlab, julia, python.");
}
```

Reply:
Even better you can add this utility code to omc_file.h/c so we can use it other places as well. I found this looking online
http://www.codebind.com/cpp-tutorial/c-get-current-directory-linuxwindows/
Seems okay to me thought we might need to modify it a bit for ourselves. 

And as far as you are concerned (if you don't mind adding 1 or 2 secs of compilation time to this file) you can just use boost::filesystem and avoid the headache.

Reply:
I did not put as much thought in it since my guess is, that being unable to get cwd results in a failing process anyway, but you are probably right.

I just let the testsuite run with my current changes to see if it works out, afterwards i will add that!

Reply:
LGTM.

Reply:
> Even better you can add this utility code to omc_file.h/c so we can use it other places as well. I found this looking online
> http://www.codebind.com/cpp-tutorial/c-get-current-directory-linuxwindows/

Good point, and then someone should try it on a Windows machine, just in case.